c/parallel/test/build_result_caching.h

```diff
@@ -0,0 +1,177 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <sstream>
+#include <tuple>
+#include <typeinfo>
+#include <unordered_map>
+
+template <typename ResultT, typename CleanupCallable>
+class result_wrapper_t
+{
+  std::shared_ptr<ResultT> m_owner;
+
+public:
+  result_wrapper_t()
+      : m_owner{}
+  {}
+  result_wrapper_t(ResultT v)
+      : m_owner{std::make_shared<ResultT>(v)}
+  {}
+
+  result_wrapper_t(const result_wrapper_t&) = default;
+  result_wrapper_t(result_wrapper_t&&)      = default;
+
+  result_wrapper_t& operator=(const result_wrapper_t&) = default;
+  result_wrapper_t& operator=(result_wrapper_t&&)      = default;
+
+  ~result_wrapper_t() noexcept
+  {
+    if (!m_owner)
+    {
+      return;
+    }
+    try
+    {
+      if (m_owner.use_count() <= 1)
+      {
+        // release resources
+        CleanupCallable{}(m_owner.get());
+      }
+    }
+    catch (const std::exception& e)
+    {
+      std::cerr << "~result_wrapper_t ignores exception: " << e.what() << std::endl;
+    }
+  }
+
+  ResultT& get()
+  {
+    return *m_owner.get();
+  }
+};
+
+template <typename KeyT, typename ValueT>
+class build_cache_t
+{
+  std::unordered_map<KeyT, ValueT> m_map;
+
+public:
+  build_cache_t()
+      : m_map{} {};
+
+  bool contains(const KeyT& key) const
+  {
+    // unorder_map::contains is C++20 feature
+    return m_map.contains(key);
+  }
+
+  void insert(const KeyT& key, ValueT&& new_value)
+  {
+    m_map[key] = std::move(new_value);
+  }
+
+  ValueT& get(const KeyT& key)
+  {
+    return m_map[key];
```

Comment:
Important: This will default construct an entry if there was none. Is that intentional?

Reply:
Not really, but `get` is only called if `m_map.contains(key)` returned `true`.

Reply:
I will add `assert(!m_map.contains(key))` to make it clear.